### PR TITLE
Added support for centralized dynamic config

### PR DESCRIPTION
**Pull request by kevsmith**
Wanted to get review on the code early while I'm working on test coverage.

### CODE REVIEW

File: example_relay.conf [hunk 1]
```diff
@@ -19,7 +19,13 @@ max_concurrent: 8
 # Environment variable: $RELAY_DYNAMIC_CONFIG_ROOT
 # Default: none
 # Required: No
-# dynamic_config_root: /tmp/bundle_configs
+# dynamic_config_root: /tmp/dynamic_configs
+
+# Use managed dynamic configuration
+# Requires dynamic_config_root
+# Environment variable: RELAY_MANAGED_DYNAMIC_CONFIG
```
Comment (christophermaier) [inline example_relay.conf]: We'll need to add this to the doc site: http://docs.operable.io/docs/relay-environment-variables

File: relay/config/config.go [hunk 2]
```diff
@@ -73,6 +76,12 @@ func (c *Config) Verify() error {
 	if c.DockerEnabled() == false && c.NativeEnabled() == false {
 		return errorNoExecutionEngines
 	}
+	if c.ManagedDynamicConfig == true && c.DynamicConfigRoot == "" {
+		return errorMissingDynamicConfigRoot
+	}
+	if c.ManagedDynamicConfig == true {
+		c.DynamicConfigRoot = path.Join(c.DynamicConfigRoot, "managed")
```
Comment (christophermaier) [inline relay/config/config.go]: This means we can't ever have a bundle named "managed", correct? We should guard against that in Cog, then (not a big deal; we already have a few other protected bundle names, and the rest of the app deals with them fine; we just need to add a new string for it to guard against)

File: relay/dynamic_config.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+			panic(err)
+		}
+	}
+}
+
+func (dcu *DynamicConfigUpdater) dynConfigUpdate(conn bus.Connection, topic string, payload []byte) {
+	defer dcu.refreshTimer.Reset(refreshInterval)
+	var envelope messages.DynamicConfigsResponseEnvelope
+	decoder := json.NewDecoder(bytes.NewReader(payload))
+	decoder.UseNumber()
```
Comment (christophermaier) [inline relay/dynamic_config.go]: One day, we should try and pull all the JSON decoding into a module so we don't have to keep repeating this same song and dance... 😦 . I'm afraid one day we're going to add something new and forget about this little wrinkle.

File: example_relay.conf [hunk 4]
```diff
@@ -19,7 +19,19 @@ max_concurrent: 8
...
+# Requires dynamic_config_root
+# Environment variable: $RELAY_MANAGED_DYNAMIC_CONFIG
+# Default: false
+# managed_dynamic_config: false
+
+# Refresh interval for managed dynamic configuration files.
+# Valid time units are s (seconds), m (minutes), and h (hours).
+# Requires managed_dynamic_config
+# Environment variable: $RELAY_MANAGED_DYNAMIC_CONFIG_INTERVAL
+# Default: 5s
```
Comment (christophermaier) [inline example_relay.conf:34]: We'll need to add these to the doc site: http://docs.operable.io/docs/relay-environment-variables

File: relay/dynamic_config.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+			panic(err)
+		}
+	}
+}
+
+func (dcu *DynamicConfigUpdater) dynConfigUpdate(conn bus.Connection, topic string, payload []byte) {
+	defer dcu.refreshTimer.Reset(dcu.refreshInterval)
+	var envelope messages.DynamicConfigsResponseEnvelope
+	decoder := json.NewDecoder(bytes.NewReader(payload))
+	decoder.UseNumber()
```
Comment (christophermaier) [inline relay/dynamic_config.go]: One day, we should try and pull all the JSON decoding into a module so we don't have to keep repeating this same song and dance... 😦 . I'm afraid one day we're going to add something new and forget about this little wrinkle.

File: relay/dynamic_config.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	if err := dcu.conn.Publish("bot/relays/info", raw); err != nil {
+		log.Errorf("Error requesting bundle dynamic configuration update: %s.", err)
+		dcu.refreshTimer.Reset(dcu.refreshInterval)
+	}
+}
+
+func (dcu *DynamicConfigUpdater) loop() {
+	<-dcu.control
+	dcu.refreshTimer.Stop()
+	dcu.conn.Disconnect()
```
Comment (christophermaier) [inline relay/dynamic_config.go:115]: How does this loop?

File: example_relay.conf (hunk 4 above)
Comment (kevsmith) [inline example_relay.conf:34]: Already done: http://docs.operable.io/v0.8/docs/relay-environment-variables-2

Will flip this page when we release 0.9.

Comment (christophermaier) [inline example_relay.conf:34]: 👍

**Comment by christophermaier** [comment]
👍

**Pull request merged**